rfd/000?-database-access.md

```diff
@@ -0,0 +1,512 @@
+$ ls
+server.cas server.crt server.key
+```
+
+Note that the `--host` parameter should match the hostname of the endpoint
+the database will be connected at.
+
+The certificate is signed by Teleport host authority which is intended for 
+machine-to-machine communication.
+
+*TODO(r0mant): Instead of using host authority, introduce another authority
+specifically for db access? Would also help decouple authority rotation for
+databases from the rest of the cluster.*
+
+The generated secrets are then used to configure the database. For example,
+for PostgreSQL:
+
+```sh
+$ cat postgresql.conf | grep ssl
+ssl = on
+ssl_ca_file = '/path/to/server.cas'
+ssl_cert_file = '/path/to/server.crt'
+ssl_key_file = '/path/to/server.key'
+```
+
+In addition, PostgreSQL access configuration file should be configured to 
+require client certificate authentication:
+
+```sh
+$ cat pg_hba.conf
+# TYPE  DATABASE        USER            ADDRESS                 METHOD
+hostssl all             all             ::/0                    cert
+hostssl all             all             0.0.0.0/0               cert
+```
+
+#### RDS/Aurora
+
+There are a few things that need to be done in order to configure RDS/Aurora
+database to support IAM authentication.
+
+1. Enable IAM authentication when provisioning a new database, or on an
+existing one. By default, only password authentication is enabled.
+
+https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.Enabling.html
+
+2. Configure IAM policy in order to allow the user Teleport database service
+will use to connect to the database instance with the auth token. An example
+policy is shown below.
+
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": "rds-db:connect",
+            "Resource": [
+                "arn:aws:rds-db:us-west-1:1234567890:dbuser:db-ABCDEFGHIJKLMNOP/developer",
+            ]
+        }
+    ]
+}
+```
+
+Note that the above rule allows IAM authentication to a single database user,
+it also supports wildcards.
+
+https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.IAMPolicy.html
+
+3. Grant `rds_iam` role to the database users. For example, for PostgreSQL:
+
+```sql
+CREATE USER developer; 
+GRANT rds_iam TO developer;
+```
+
+https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.DBAccounts.html
+
+## RBAC
+
+Teleport role resource gets 3 new fields that allow to control access to 
+database instances as well as individual databases and database users.
+
+```yaml
+kind: role
+version: v3
+metadata:
+  name: developer
+spec:
+  allow:
+    # Label selectors for database instances this role has access to.
+    db_labels:
+      environment: ["dev", "stage"]
+    # Database names (within a database instance) this role has access to.
+    db_names: ["main", "metrics", "postgres"]
+    # Database users this role can log in as.
+    db_users: ["alice", "bob"]
```

Comment:
how does this map to IAM users?

Reply:
These users are enforced at the protocol level (so it works both for onprem and RDS). IAM-wise, like shown in the example above, you define a IAM policy that allows access as certain database users. For example, in this case a policy might look like this:

```yaml
        {
            "Effect": "Allow",
            "Action": "rds-db:connect",
            "Resource": [
                "arn:aws:rds-db:us-west-1:1234567890:dbuser:db-ABCDEFGHIJKLMNOP/alice",
                "arn:aws:rds-db:us-west-1:1234567890:dbuser:db-ABCDEFGHIJKLMNOP/bob",
            ]
        }
```

And then attached to the IAM user that database service uses.

Reply:
can I define `arn:aws:rds-db:us-west-1:1234567890:dbuser:db-ABCDEFGHIJKLMNOP/*`?

I think you should add this to the document to explain people

Reply:
Yes, it supports wildcards. It's actually already mentioned in the RFD, but you're right it's not very clear, I'll make this wildcard example more prominent.